src/realm/array_direct.hpp

```diff
@@ -200,6 +431,560 @@ inline std::pair<int64_t, int64_t> get_two(const char* data, size_t width, size_
     REALM_TEMPEX(return get_two, width, (data, ndx));
 }
 
+/* Subword parallel search
+
+    The following provides facilities for subword parallel search for bitfields of any size.
+    To simplify, the first bitfield must be aligned within the word: it must occupy the lowest
+    bits of the word.
+
+    In general the metods here return a vector with the most significant bit in each field
+    marking that a condition was met when comparing the corresponding pair of fields in two
+    vectors. Checking if any field meets a condition is as simple as comparing the return
+    vector against 0. Finding the first to meet a condition is also supported.
+
+    Vectors are "split" into fields according to a MSB vector, wich indicates the most
+    significant bit of each field. The MSB must be passed in as an argument to most
+    bit field comparison functions. It can be generated by the field_sign_bit<width> template.
+
+    The simplest condition to test is any_field_NE(A,B), where A and B are words.
+    This condition should be true if any bitfield in A is not equal to the corresponding
+    field in B.
+
+    This is almost as simple as a direct word compare, but needs to take into account that
+    we may want to have part of the words undefined.
+*/
+constexpr int num_fields_table[65] = {-1, 64, 32, 21, 16, 12, 10, 9, // 0-7
+                                      8,  7,  6,  5,  5,  4,  4,  4, // 8-15
+                                      4,  3,  3,  3,  3,  3,  2,  2, // 16-23
+                                      2,  2,  2,  2,  2,  2,  2,  2, // 24-31
+                                      2,  1,  1,  1,  1,  1,  1,  1, // 32-39
+                                      1,  1,  1,  1,  1,  1,  1,  1, // 40-47
+                                      1,  1,  1,  1,  1,  1,  1,  1, // 48-55
+                                      1,  1,  1,  1,  1,  1,  1,  1, // 56-63
+                                      1};
+
+constexpr int num_bits_table[65] = {-1, 64, 64, 63, 64, 60, 60, 63, // 0-7
+                                    64, 63, 60, 55, 60, 52, 56, 60, // 8-15
+                                    64, 51, 54, 57, 60, 63, 44, 46, // 16-23
+                                    48, 50, 52, 54, 56, 58, 60, 64, // 24-31
+                                    64, 33, 34, 35, 36, 37, 38, 39, // 32-39
+                                    40, 41, 42, 43, 44, 45, 46, 47, // 40-47
+                                    48, 49, 50, 51, 52, 53, 54, 55, // 48-55
+                                    56, 57, 58, 59, 60, 61, 62, 63, // 56-63
+                                    64};
+
+inline int num_fields_for_width(size_t width)
```

Comment:
Yep, I think it does not make sense, but in release mode the assertions are disabled right? So the reason for that is this:
constexpr int num_fields_table[65] = {-1,

Reply:
I haven't changed the return type, mainly because of this: 
```
template <typename VectorCompare>
size_t parallel_subword_find(VectorCompare vector_compare, const uint64_t* data, size_t offset, size_t width,
                             uint64_t MSBs, uint64_t search_vector, size_t start, size_t end)
{
    const auto field_count = num_fields_for_width(width);
```
Which basically stops the search in case for some reason we have width set to 0. Now, during the compression of the array, we are basically asserting that width must not be 0. So probably we could just throw there, and change all these APIs later.  But maybe it is better to do it once, the whole code is reviewed and merged.  What do you think?

Reply:
Actually scratch this, I have fixed the return type but also the type of `num_bits_table` and the value returned for width = 0, which was -1. Which gave us ubsan problems, since:
`0xFFFFFFFFFFFFFFFFULL >> (64 - num_bits_table[0]);` == `0xFFFFFFFFFFFFFFFFULL >> 65`, which is obviously wrong.